llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,320 @@
+    }
+    Size = NewSize;
+    // If the capacity is enough, just update the size and continue
+    // with the currently allocated pages. Notice that we do not
+    // need to default construct any new element, because that was already done
+    // when the page was allocated.
+    if (Size <= capacity())
+      return;
+    // The number of pages to allocate. The Remainder is calculated
+    // for the case in which the NewSize is not a multiple of PageSize.
+    // In that case we need one more page.
+    size_t Pages = Size / PageSize;
+    size_t Remainder = Size % PageSize;
+    if (Remainder != 0)
+      Pages += 1;
+    assert(Pages > PageToDataPtrs.size());
+    // We use InvalidPage to indicate that a page has not been allocated yet.
+    // This cannot be 0, because 0 is a valid page id.
+    // We use InvalidPage instead of a separate bool to avoid wasting space.
+    PageToDataPtrs.resize(Pages, InvalidPage);
+  }
+
+  [[nodiscard]] bool empty() const { return Size == 0; }
+
+  /// Clear the vector, i.e. clear the allocated pages, the whole page
+  /// lookup index and reset the size.
+  void clear() {
+    Size = 0;
+    for (T *Page : PageToDataPtrs) {
+      if (Page == InvalidPage)
+        continue;
+      std::destroy_n(Page, PageSize);
+      // If we do not own the allocator, deallocate the pages one by one.
+      if (!Allocator.getInt())
+        Allocator.getPointer()->Deallocate(Page);
+    }
+    // If we own the allocator, simply reset it.
+    if (Allocator.getInt() == true)
+      Allocator.getPointer()->Reset();
+    PageToDataPtrs.clear();
+  }
+
+  /// Iterator on all the elements of the vector
+  /// which have actually being constructed.
+  class MaterialisedIterator {
+    const PagedVector *PV;
+    size_t ElementIdx;
+
+  public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = T;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T *;
+    using reference = T &;
+
+    MaterialisedIterator(PagedVector const *PV, size_t ElementIdx)
+        : PV(PV), ElementIdx(ElementIdx) {}
+
+    /// Pre-increment operator.
+    ///
+    /// When incrementing the iterator, we skip the elements which have not
+    /// been materialised yet.
+    MaterialisedIterator &operator++() {
+      ++ElementIdx;
+      if (ElementIdx % PageSize == 0) {
+        while (ElementIdx < PV->Size &&
+               PV->PageToDataPtrs[ElementIdx / PageSize] == InvalidPage)
+          ElementIdx += PageSize;
+        if (ElementIdx > PV->Size)
+          ElementIdx = PV->Size;
+      }
+
+      return *this;
+    }
+
+    MaterialisedIterator operator++(int) {
+      MaterialisedIterator Copy = *this;
+      ++*this;
+      return Copy;
+    }
+
+    /// Dereference operator.
+    ///
+    /// Notice this can materialise elements if needed so there might be
+    /// a page allocation and additional construction of the elements of
+    /// such page.
```

Comment:
This does not appear to be true.
```suggestion
```

Reply:
I think this is still true, no? operator* invokes operator[] which will allocate pages and construct elements, as needed.

Reply:
No. This does not invoke operator[] on the PagedVector, and in any case a MaterialisedIterator never points to an unmaterialized element.

Reply:
Indeed, sorry for misreading. I will drop the comment.